src/search/task_proxy.h

```diff
@@ -558,84 +551,71 @@ class GoalsProxy : public ConditionsProxy {
 
-    bool operator==(const State &other) const {
-        assert(task == other.task);
-        return values == other.values;
-    }
+    /* Generate unpacked data if none is available yet. Calling the function
+       on a state that already has unpacked has no effect. */
+    void unpack() const;
 
-    bool operator!=(const State &other) const {
-        return !(*this == other);
-    }
+    std::size_t size() const;
+    FactProxy operator[](std::size_t var_id) const;
+    FactProxy operator[](VariableProxy var) const;
 
-    std::size_t size() const {
-        return values.size();
-    }
+    inline TaskProxy get_task() const;
 
-    FactProxy operator[](std::size_t var_id) const {
-        assert(var_id < size());
-        return FactProxy(*task, var_id, values[var_id]);
-    }
+    /* Return a pointer to the registry in which this state is registered.
+       If the state is not registered, return nullptr. */
+    const StateRegistry *get_registry() const;
+    /* Return the ID of the state within its registry. If the state is not
+       registered, return StateID::no_state. */
+    StateID get_id() const;
 
-    FactProxy operator[](VariableProxy var) const {
-        return (*this)[var.get_id()];
-    }
+    /* Access the unpacked values. Accessing the unpacked values in a state
+       that that doesn't have them is an error. Use unpack() to ensure the
+       data exists. */
+    const std::vector<int> &get_unpacked_values() const;
 
-    inline TaskProxy get_task() const;
-
-    const std::vector<int> &get_values() const {
-        return values;
-    }
+    /* Access the packed values. Accessing packed values on states that do
+       not have them (unregistered states) is an error. */
+    const PackedStateBin *get_buffer() const;
 
-    State get_successor(OperatorProxy op) const {
-        if (task->get_num_axioms() > 0) {
-            ABORT("State::get_successor currently does not support axioms.");
-        }
-        assert(!op.is_axiom());
-        //assert(is_applicable(op, state));
-        std::vector<int> new_values = values;
-        for (EffectProxy effect : op.get_effects()) {
-            if (does_fire(effect, *this)) {
-                FactProxy effect_fact = effect.get_fact();
-                new_values[effect_fact.get_variable().get_id()] = effect_fact.get_value();
-            }
-        }
-        return State(*task, std::move(new_values));
-    }
+    /*
+      Create a successor state with the given operator. The operator is assumed
+      to be applicable and the precondition is not checked. This will create an
+      unpacked, unregistered successor. If you need registered successors, use
+      the methods of StateRegistry.
+    */
+    State get_unregistered_successor(const OperatorProxy &op) const;
 };
 
 
 namespace utils {
 inline void feed(HashState &hash_state, const State &state) {
-    feed(hash_state, state.get_values());
+    feed(hash_state, state.get_unpacked_values());
```

Comment:
This will fail for packed-only states, but maybe it should?

Reply:
Good spot. It isn't intentional, so we could change it. Storing packed states in a hash set does not seem like something you shouldn't do. Is there a feed method where we can pass in just a piece of memory (buffer + length)?

Reply:
Do we use this for the state registry, or does this use another hashing path? I'm not aware of anywhere where we hash unregistered states. For registered states, using the packed representation for the hashes would be preferable because it is smaller. (In the future, we may also want to look at incremental hashing, for example using Zobrist hash functions.)

Reply:
The state registry uses its own way of computing hashes because we don't actually add states to the hash set there but indices into the data structure storing the state data. This function here is used in diverse potential heuristics in a map that maps sample states to optimal potential functions for them.

Reply:
We discussed this a bit more. We (currently, this could be up for discussion) don't want to unpack implicitly, so this rules out the option of unpacking the states here on demand. If we don't want to unpack on demand, the most obvious option to support both packed and unpacked states (of which the latter don't always have packed data) is to sometimes use the packed and sometimes use the unpacked data. (There is also the option of computing the hash of the unpacked data for packed states on demand without physically unpacking, but that seems bizarre.)

But mixing both kinds of hashes also causes problem. We could then have two logically equal states that hash differently. I suggest that for now we forbid hashing states without an unpacked representation and add an assertion. Hashing packed (and therefore registered) states also seems like a performance error because it's much cheaper to hash the state IDs instead.

Reply:
We probably want to explain this (maybe a bit shorter) in the code, too.

Reply:
We cannot test this with an assertion because there is no way publicly available way to test if a state has unpacked data available. But even without an assertion, the code will fail with an error message from within `get_unpacked_values` in this case. I added a comment to explain things.